### PR TITLE
Split type declaration into its own page

### PR DESCRIPTION
**Pull request by Girgias**
Also adds a section about union types.

### CODE REVIEW

File: language/types/declarations.xml [hunk 1]
```diff
@@ -0,0 +1,754 @@
...
+     <row>
+      <entry><literal>mixed</literal></entry>
+      <entry>
+       The value can be any value.
+      </entry>
+      <entry>PHP 8.0.0</entry>
+     </row>
+    </tbody>
+   </tgroup>
+  </informaltable>
```
Comment (cmb69) [inline language/types/declarations.xml:129]: Why are some of the types in this table marked up as `<type>` and others as `<literal>`? Doesn't look reasonable to me. (and yes, I'm aware that this table has just been moved)

File: language/types/declarations.xml [hunk 2]
```diff
@@ -0,0 +1,754 @@
...
+   <title>false pseudo-type</title>
+   <para>
+    The <literal>false</literal> literal type is supported as part of unions,
+    and is included as for historical reasons many internal functions return
+    <literal>false</literal> instead of <literal>null</literal> for failures.
+    A classic example of such a function is <function>strpos</function>.
+    such that <literal>T1|T2|null</literal> can be used to create a nullable union.
+    The existing <literal>?T</literal> notation is considered a shorthand
+    for the common case of <literal>T|null</literal>.
+   </para>
```
Comment (cmb69) [inline language/types/declarations.xml:370]: The 3 last lines appears to superfluous (copy-paste issue?)

File: language/types/declarations.xml [hunk 3]
```diff
@@ -0,0 +1,754 @@
...
+     As such, all of <literal>false</literal>, <literal>false|null</literal>
+     and <literal>?false</literal> are not permitted.
+    </simpara>
+   </warning>
+   <warning>
+    <simpara>
+     The <literal>true</literal> literal type does <emphasis>not</emphasis>
+     exist.
+    </simpara>
+   </warning>
```
Comment (cmb69) [inline language/types/declarations.xml:385]: Maybe `<warning>` is a bit too strong for both cases?

File: language/types/declarations.xml [hunk 4]
```diff
@@ -0,0 +1,754 @@
...
+    <itemizedlist>
+     <listitem>
+      <simpara>
+       Each name-resolved type may only occur once. Types such as
+       <literal>int|string|INT</literal> result in an error.
+      </simpara>
+     </listitem>
+     <listitem>
+      <simpara>
+       If <literal>bool</literal> is used, <literal>false</literal> cannot be used additionally.
```
Comment (cmb69) [inline language/types/declarations.xml:404]:
```suggestion
       If <type>bool</type> is used, <type>false</type> cannot be used additionally.
```
PhD support for `false` should be added explicitly, but this should work reasonably well right now.

**Review by cmb69 — COMMENTED** [review]
Thanks for working on this! Moving the type declaration documentation to an appropriate place is long overdue, and having union types documented help to simplify the [PHP 8.0 migration guide](https://github.com/php/doc-en/pull/170) a bit.

File: language/types/declarations.xml [hunk 5]
```diff
@@ -0,0 +1,754 @@
...
+    </informalexample>
+   </para>
+  </sect3>
+
+ </sect2>
+
+ <sect2 xml:id="language.types.declarations.return-only">
+  <title>Return only types</title>
+
+  <sect3 xml:id="language.types.declarations.void">
```
Comment (cmb69) [inline language/types/declarations.xml:461]: Would it make sense to somehow unify/merge this section with https://github.com/php/doc-en/blob/cdaea0421544885f02ff3d36bd203dc01b78299e/language/types/pseudo-types.xml#L34-L43

and have `void`'s own section (`&language.types.void;`)? I'd really like to drop the pseudo-types section; `mixed` is a built-in type as well now, and `number` could be replaced with the respective union type; and variadics are being worked on already.

**Comment by cmb69** [comment]
Also, before actually removing a page from the manual ([language/oop5/typehinting.xml](https://www.php.net/manual/en/language.oop5.typehinting.php) in this case), it might be sensible to cater to existing user notes (i.e. integrate or remove).

File: language/types/declarations.xml (hunk 2 above)
Comment (Girgias) [reply language/types/declarations.xml:370]: I indeed copy pasted this from the Union type RFC, don't know if it's superfluous, but might make more sense as a note.

File: language/types/declarations.xml (hunk 3 above)
Comment (Girgias) [reply language/types/declarations.xml:385]: Should I use ``<caution>`` instead?

File: language/types/declarations.xml (hunk 5 above)
Comment (Girgias) [reply language/types/declarations.xml:461]: Seems reasonable, I suppose I could add a note which goes along the lines:
```xml
<note>
 <para>
   The PHP documentation uses <literal>void</literal> in an argument list to convey that a function doesn't accept any arguments.
 </para>
</note>
```

**Comment by Girgias** [comment]
> 
> 
> Also, before actually removing a page from the manual ([language/oop5/typehinting.xml](https://www.php.net/manual/en/language.oop5.typehinting.php) in this case), it might be sensible to cater to existing user notes (i.e. integrate or remove).

The notes all seems updated and irrelevant to some extend by just being benchmarks (I noted that Nikita also removed a bunch just now)

File: language/types/declarations.xml (hunk 3 above)
Comment (cmb69) [reply language/types/declarations.xml:385]: IMO, yes. :)

File: language/types/declarations.xml [hunk 6]
```diff
@@ -0,0 +1,817 @@
...
+  <para>
+   Type declarations were also known as type hints in previous versions of PHP.
+  </para>
+ </note>
+
+ <para>
+  To specify a type declaration, the type name should be added before the
+  parameter name. The declaration can be made to accept &null; values if
+  the default value of the parameter is set to &null;.
+ </para>
```
Comment (Girgias) [inline language/types/declarations.xml:23]: This should probably dropped now and/or integrated in some other paragraph

File: language/types/declarations.xml [hunk 7]
```diff
@@ -0,0 +1,817 @@
...
+    function test(boolean $param) {}
+    test(true);
+?>
+]]>
+     </programlisting>
+     &example.outputs;
+     <screen>
+<![CDATA[
+    Fatal error: Uncaught TypeError: Argument 1 passed to test() must be an instance of boolean, boolean given, called in - on line 1 and defined in -:1
+]]>
```
Comment (Girgias) [inline language/types/declarations.xml:154]: I think the test output needs to be updated, as I think it now says "bool given"

File: language/types/declarations.xml [hunk 8]
```diff
@@ -0,0 +1,817 @@
...
+    such that <literal>T1|T2|null</literal> can be used to create a nullable union.
+    The existing <literal>?T</literal> notation is considered a shorthand
+    for the common case of <literal>T|null</literal>.
+   </para>
+   
+   <warning>
+    <simpara>
+     <literal>null</literal> cannot be used as a standalone type.
+    </simpara>
+   </warning>
```
Comment (Girgias) [inline language/types/declarations.xml:390]: Should I mark this as caution too now that I made `false` a caution section?

File: language/types/declarations.xml [hunk 9]
```diff
@@ -0,0 +1,817 @@
...
+  </sect3>
+ </sect2>
+
+ <sect2 xml:id="language.types.declarations.strict">
+  <title>Strict typing</title>
+
+  <para>
+   By default, PHP will coerce values of the wrong type into the expected
+   scalar type declaration if possible. For example, a function that is given
+   an <type>integer</type> for a parameter that expects a <type>string</type>
```
Comment (Girgias) [inline language/types/declarations.xml:520]:
```suggestion
   an <type>int</type> for a parameter that expects a <type>string</type>
```
Should this use the canonical form or not?

File: language/types/declarations.xml (hunk 8 above)
Comment (cmb69) [reply language/types/declarations.xml:390]: IMO, yes.

File: language/types/declarations.xml (hunk 9 above)
Comment (cmb69) [reply language/types/declarations.xml:520]: IMO, yes.

File: language/types/declarations.xml (hunk 5 above)
Comment (nikic) [reply language/types/declarations.xml:494]: Maybe we should drop that as well? Using `void` to indicate no arguments is a C-ism, I don't think there's any value in doing that in the PHP manual.

Comment (cmb69) [reply language/types/declarations.xml:494]: See https://github.com/php/phd/pull/32

File: language/types/declarations.xml [hunk 10]
```diff
@@ -0,0 +1,817 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- $Revision$ -->
+<sect1 xml:id="language.types.declarations">
+ <title>Type declarations</title>
+
+ <para>
+  Type declarations can be added to function arguments, return values,
+  and, as of PHP 7.4.0, class properties. And they ensure that the value
```
Comment (nikic) [inline language/types/declarations.xml:8]:
```suggestion
  and, as of PHP 7.4.0, class properties. They ensure that the value
```

File: language/types/declarations.xml [hunk 11]
```diff
@@ -0,0 +1,817 @@
...
+  is of the specified type at call time, otherwise a
+  <classname>TypeError</classname> is thrown.
+ </para>
+
+ <note>
+  <para>
+   Type declarations were also known as type hints in previous versions of PHP.
+  </para>
```
Comment (nikic) [inline language/types/declarations.xml:16]: Let's make Levi happy and drop this sentence :)

Comment (cmb69) [reply language/types/declarations.xml:16]: Or may be add that the term "type hints" is frowned upon?

File: language/types/declarations.xml [hunk 12]
```diff
@@ -0,0 +1,819 @@
...
+]]>
+    </programlisting>
+   </example>
+  </para>
+
+  <note>
+   <para>
+    Prior to PHP 7.1.0, it was possible to achieve nullable arguments by making
+    <literal>null</literal> the default value.
+    This is not recommended as this breaks during inheritance.
```
Comment (Girgias) [inline language/types/declarations.xml:338]: Should an example about this be added?

File: language/types/declarations.xml [hunk 13]
```diff
@@ -0,0 +1,819 @@
...
+42.1  --> 42          // float compatible with int
+1e100 --> "1.0E+100"  // float too large for int type, fall back to string
+INF   --> "INF"       // float too large for int type, fall back to string
+true  --> 1           // bool compatible with int
+[]    --> TypeError   // array not compatible with int or string
+
+// int|float|bool
+"45"    --> 45        // int numeric string
+"45.0"  --> 45.0      // float numeric string
+"45X"   --> 45 + Notice: Non well formed numeric string
```
Comment (Girgias) [inline language/types/declarations.xml:719]: That wouldn't work any more right @nikic ?

Maybe this should be made into "proper" examples

**Comment by Girgias** [comment]
So I still don't know what to do with the two examples in the temporary "Misc" section, any ideas?

File: language/types/declarations.xml (hunk 6 above)
Comment (cmb69) [reply language/types/declarations.xml:17]: Yes, could go to language.types.declarations.nullable.

File: language/types/declarations.xml (hunk 12 above)
Comment (cmb69) [reply language/types/declarations.xml:338]: I don't think that's necessary.

File: language/types/declarations.xml (hunk 13 above)
Comment (cmb69) [reply language/types/declarations.xml:718]: "Proper" examples (using example.outputs and friends) can be hard to read for a lot of output. Maybe split out the cases which have different behavior for different PHP versions?

File: language/types/declarations.xml [hunk 14]
```diff
@@ -0,0 +1,818 @@
...
+]]>
+    </programlisting>
+    &example.outputs;
+    <screen>
+<![CDATA[
+int(3)
+Error: sum(): Argument #1 ($a) must be of type int, float given, called in - on line 10
+]]>
+    </screen>
+   </example>
```
Comment (cmb69) [inline language/types/declarations.xml:794]: These two examples are PHP 8 only; at least that should be mentioned. Also use `&example.outputs.80;` (needs to be added to language-snippets.ent).

File: language/types/declarations.xml (hunk 13 above)
Comment (Girgias) [reply language/types/declarations.xml:718]: Well as it's about union types it can only be PHP 8.0.

Comment (cmb69) [reply language/types/declarations.xml:718]: Good point!

File: language/types/declarations.xml [hunk 15]
```diff
@@ -0,0 +1,812 @@
...
+   type under PHPs existing type checking semantics, then the type is chosen.
+   Otherwise the next type is tried.
+  </para>
+
+  <caution>
+   <para>
+    As an exception, if the value is a string and both int and float are part
+    of the union, the preferred type is determined by the existing
+    “numeric string” semantics.
+    For example, for <literal>“42”</literal> <type>int</type> is chosen,
```
Comment (cmb69) [inline language/types/declarations.xml:679]: Not quite sure about using proper quotes in the text, but the `“42”` is definitely misleading.

Comment (Girgias) [reply language/types/declarations.xml:679]: Oh yeah I don't know how that slipped in (probably from the copy and paste). I'll fix this before merging.

**Pull request closed without merging**